src/sensors/bmi160sensor.cpp

```diff
@@ -110,110 +140,525 @@ void BMI160Sensor::motionSetup() {
+                
+                constexpr float invPeriod = 1.0f / BMI160_ODR_GYR_MICROS;
+                int32_t sampleOffset = round((float)dtMicros * invPeriod) - 1;
+                if (abs(sampleOffset) > 3) {
+                    dtMicros = sampleDtMicros;
+                } else if (sampleOffset != 0) {
+                    dtMicros -= sampleOffset * sampleDtMicros;
+                }
+
+                onGyroRawSample(dtMicros, gx, gy, gz);
+            }
+        } else {
+            break;
+        }
     }
-#endif
+}
+
+void BMI160Sensor::onGyroRawSample(uint32_t dtMicros, int16_t x, int16_t y, int16_t z) {
+    #if BMI160_DEBUG
+        gyrReads++;
+    #endif
 
-    if (!OPTIMIZE_UPDATES || !lastQuatSent.equalsWithEpsilon(quaternion))
+    #if BMI160_USE_TEMPCAL
+        bool restDetected = false;
+        #if BMI160_VQF_REST_DETECTION_AVAILABLE
+            restDetected = vqf.getRestDetected();
+        #else
+            restDetected = restDetection.getRestDetected();
+        #endif
+        gyroTempCalibrator->updateGyroTemperatureCalibration(temperature, restDetected, x, y, z);
+        if (gyroTempCalibrator->shouldSaveConfig()) {
+            gyroTempCalibrator->saveConfig();
+        }
+    #endif
+
+    sensor_real_t gyroCalibratedStatic[3];
+    gyroCalibratedStatic[0] = (sensor_real_t)((((double)x - m_Calibration.G_off[0]) * gscaleX));
+    gyroCalibratedStatic[1] = (sensor_real_t)((((double)y - m_Calibration.G_off[1]) * gscaleY));
+    gyroCalibratedStatic[2] = (sensor_real_t)((((double)z - m_Calibration.G_off[2]) * gscaleZ));
+
+    #if BMI160_USE_TEMPCAL
+    float GOxyz[3];
+    if (gyroTempCalibrator->approximateOffset(temperature, GOxyz)) {
+        Gxyz[0] = (sensor_real_t)((((double)x - GOxyz[0] - GOxyzStaticTempCompensated[0]) * gscaleX));
+        Gxyz[1] = (sensor_real_t)((((double)y - GOxyz[1] - GOxyzStaticTempCompensated[1]) * gscaleY));
+        Gxyz[2] = (sensor_real_t)((((double)z - GOxyz[2] - GOxyzStaticTempCompensated[2]) * gscaleZ));
+    }
+    else
+    #endif
     {
-        newData = true;
-        lastQuatSent = quaternion;
+        Gxyz[0] = gyroCalibratedStatic[0];
+        Gxyz[1] = gyroCalibratedStatic[1];
+        Gxyz[2] = gyroCalibratedStatic[2];
     }
+    remapGyroAccel(&Gxyz[0], &Gxyz[1], &Gxyz[2]);
+
+    #if !BMI160_VQF_REST_DETECTION_AVAILABLE
+        restDetection.updateGyr(dtMicros, Gxyz);
+    #endif
+    #if USE_6_AXIS
+        #if BMI160_USE_VQF
+            vqf.updateGyr(Gxyz, (double)dtMicros * 1.0e-6);
+        #else
+            mahony.updateInto(qwxyz, Axyz[0], Axyz[1], Axyz[2], Gxyz[0], Gxyz[1], Gxyz[2], ((float)dtMicros) * 1.0e-6);
+        #endif
+    #else
+        #if BMI160_USE_VQF
+            vqf.updateGyr(Gxyz, (double)dtMicros * 1.0e-6);
+        #else
+            mahonyQuaternionUpdate(qwxyz, Axyz[0], Axyz[1], Axyz[2], Gxyz[0], Gxyz[1], Gxyz[2], Mxyz[0], Mxyz[1], Mxyz[2], ((float)dtMicros) * 1.0e-6f);
```

Comment:
At this point, we have 3 versions of mahony fusion. Are you sure you need C++ and C version side by side?

Reply:
There's a problem with multiple instances of pure C version due to use of static global variables - it might cause weird behavior with aux trackers, ideally it should be replaced with the C++ class everywhere.